README
Clutter-Vlc README
========================

Clutter-Vlc (clutter-vlc) is an integration library for using Vlc
with Clutter.

Clutter is an open source software library for creating fast, visually
rich and animated graphical user interfaces. It uses OpenGL for drawing
primitives and has multiple backends, allowing its usage on different
platforms.



Clutter-Vlc requires:

  GLib >= 2.10.0
  Clutter >= 0.8.0
  Vlc >= 0.9.8

Released under the terms of the GNU Lesser General Public License, version 2.1
or later. See the file COPYING for details.



NEWS
clutter-vlc 0.9.0
=================
* first version


examples/video-player.c
#include <clutter/clutter.h>
#include <clutter-vlc/clutter-vlc.h>

static ClutterColor color_light_blue = { 0xaa, 0xaa, 0xff, 0xff };

int
main(int ac, char** av)
{
  ClutterActor* stage;
  ClutterActor* texture;

  if (ac < 2)
    return -1;

  clutter_init(&ac, &av);
  clutter_vlc_init();

  stage = clutter_stage_get_default();
  clutter_stage_set_color(CLUTTER_STAGE(stage), &color_light_blue);

  texture = clutter_vlc_video_texture_new();
  clutter_media_set_uri(CLUTTER_MEDIA(texture), av[1]);
  clutter_media_set_playing(CLUTTER_MEDIA(texture), TRUE);

  clutter_actor_set_size(texture, 320, 240);
  clutter_actor_set_position(texture,
			     clutter_actor_get_width(stage) / 2 - clutter_actor_get_width(texture) / 2,
			     clutter_actor_get_height(stage) / 2 - clutter_actor_get_height(texture) / 2);

  clutter_container_add_actor(CLUTTER_CONTAINER(stage), texture);

  clutter_actor_show(stage);

  clutter_main();

  clutter_vlc_cleanup();

  return 0;
}


clutter-vlc/clutter-vlc.h
/*
 * Clutter-Vlc.
 *
 * VLC integration library for Clutter.
 *
 * Authored By Arnaud Vallat  <rno.rno@gmail.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef _HAVE_CLUTTER_VLC_H
# define _HAVE_CLUTTER_VLC_H

# include <clutter/clutter.h>

# include "clutter-vlc-main.h"
# include "clutter-vlc-video-texture.h"
# include "clutter-vlc-version.h"

#endif /* _HAVE_CLUTTER_VLC_H */


clutter-vlc/clutter-vlc-main.c
/*
 * Clutter-Vlc.
 *
 * VLC integration library for Clutter.
 *
 * Authored By Arnaud Vallat  <rno.rno@gmail.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include <clutter/clutter.h>
#include <glib.h>


void
clutter_vlc_init(void)
{
  g_thread_init(NULL);
  clutter_threads_init();
}


void
clutter_vlc_cleanup(void)
{
  /* nothing for now
   */
}


clutter-vlc/clutter-vlc-main.h
/*
 * Clutter-Vlc.
 *
 * VLC integration library for Clutter.
 *
 * Authored By Arnaud Vallat  <rno.rno@gmail.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef _HAVE_CLUTTER_VLC_MAIN_H
# define _HAVE_CLUTTER_VLC_MAIN_H


void	clutter_vlc_init(void);
void	clutter_vlc_cleanup(void);


#endif /* _HAVE_CLUTTER_VLC_MAIN_H */


clutter-vlc/clutter-vlc-video-texture.h
/*
 * Clutter-Vlc.
 *
 * VLC integration library for Clutter.
 *
 * Authored By Arnaud Vallat  <rno.rno@gmail.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#ifndef _HAVE_CLUTTER_VLC_VIDEO_TEXTURE_H
# define _HAVE_CLUTTER_VLC_VIDEO_TEXTURE_H

# include <glib-object.h>
# include <clutter/clutter.h>

G_BEGIN_DECLS

#define CLUTTER_VLC_TYPE_VIDEO_TEXTURE clutter_vlc_video_texture_get_type()

#define CLUTTER_VLC_VIDEO_TEXTURE(obj)					\
  (G_TYPE_CHECK_INSTANCE_CAST ((obj),					\
			       CLUTTER_VLC_TYPE_VIDEO_TEXTURE, ClutterVlcVideoTexture))

#define CLUTTER_VLC_VIDEO_TEXTURE_CLASS(klass)				\
  (G_TYPE_CHECK_CLASS_CAST ((klass),					\
			    CLUTTER_VLC_TYPE_VIDEO_TEXTURE, ClutterVlcVideoTextureClass))

#define CLUTTER_VLC_IS_VIDEO_TEXTURE(obj)			\
  (G_TYPE_CHECK_INSTANCE_TYPE ((obj),				\
			       CLUTTER_VLC_TYPE_VIDEO_TEXTURE))

#define CLUTTER_VLC_IS_VIDEO_TEXTURE_CLASS(klass)		\
  (G_TYPE_CHECK_CLASS_TYPE ((klass),				\
			    CLUTTER_VLC_TYPE_VIDEO_TEXTURE))

#define CLUTTER_VLC_VIDEO_TEXTURE_GET_CLASS(obj)			\
  (G_TYPE_INSTANCE_GET_CLASS ((obj),					\
			      CLUTTER_VLC_TYPE_VIDEO_TEXTURE, ClutterVlcVideoTextureClass))

typedef struct _ClutterVlcVideoTexture        ClutterVlcVideoTexture;
typedef struct _ClutterVlcVideoTextureClass   ClutterVlcVideoTextureClass;
typedef struct _ClutterVlcVideoTexturePrivate ClutterVlcVideoTexturePrivate;


struct _ClutterVlcVideoTexture
{
  ClutterTexture parent;

  ClutterVlcVideoTexturePrivate* priv;
};


struct _ClutterVlcVideoTextureClass
{
  ClutterTextureClass parent_class;
};


GType         clutter_vlc_video_texture_get_type(void) G_GNUC_CONST;
ClutterActor* clutter_vlc_video_texture_new(void);


G_END_DECLS

#endif /* _HAVE_CLUTTER_VLC_VIDEO_TEXTURE_H */


clutter-vlc/clutter-vlc-video-texture.c
/*
 * Clutter-Vlc.
 *
 * VLC integration library for Clutter.
 *
 * Authored By Arnaud Vallat  <rno.rno@gmail.com>
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with this library; if not, write to the
 * Free Software Foundation, Inc., 59 Temple Place - Suite 330,
 * Boston, MA 02111-1307, USA.
 */

#include "clutter-vlc-video-texture.h"
#include <vlc/vlc.h>

#include <stdlib.h>

#define TICK_TIMEOUT 0.5

struct _ClutterVlcVideoTexturePrivate
{
  libvlc_exception_t vlc_exception;
  libvlc_instance_t* vlc_instance;
  libvlc_media_player_t* vlc_media_player;

  gchar* uri;
  guint tick_timeout_id;
};

enum {
  PROP_0,

  /* ClutterMedia proprs */
  PROP_URI,
  PROP_PLAYING,
  PROP_PROGRESS,
  PROP_AUDIO_VOLUME,
  PROP_CAN_SEEK,
  PROP_BUFFER_FILL,
  PROP_DURATION
};


#define CLUTTER_VLC_VIDEO_TEXTURE_GET_PRIVATE(obj)			\
  (G_TYPE_INSTANCE_GET_PRIVATE((obj),					\
			       CLUTTER_VLC_TYPE_VIDEO_TEXTURE,		\
			       ClutterVlcVideoTexturePrivate))


static void clutter_vlc_media_init(ClutterMediaIface *iface);

G_DEFINE_TYPE_WITH_CODE(ClutterVlcVideoTexture,
			clutter_vlc_video_texture,
			CLUTTER_TYPE_TEXTURE,
			G_IMPLEMENT_INTERFACE(CLUTTER_TYPE_MEDIA,
					      clutter_vlc_media_init));

/* Interface implementation */

static void
clutter_vlc_catch(libvlc_exception_t* vlc_exception)
{
  /* well this is really bad. Find out another way to handle vlc exceptions
   */

  if (libvlc_exception_raised(vlc_exception))
    {
      fprintf(stderr, "exception: %s\n", libvlc_exception_get_message(vlc_exception));
      exit(1);
    }

  libvlc_exception_clear(vlc_exception);
}


static gboolean
clutter_vlc_tick_timeout(gpointer data)
{
  GObject* self;

  self = data;

  g_object_notify(self, "progress");

  return TRUE;
}


static void
clutter_vlc_set_uri(ClutterVlcVideoTexture* video_texture,
		    const gchar* uri)
{
  ClutterVlcVideoTexturePrivate* priv;
  GObject* self;
  libvlc_media_t* vlc_media;
  char media_arg[96];

  priv = video_texture->priv;
  self = G_OBJECT(video_texture);

  if (priv->vlc_instance == NULL)
    return;

  if (priv->vlc_media_player != NULL)
    {
      libvlc_media_player_stop(priv->vlc_media_player, &priv->vlc_exception);
      clutter_vlc_catch(&priv->vlc_exception);
      libvlc_media_player_release(priv->vlc_media_player);
      priv->vlc_media_player = NULL;
      g_source_remove(priv->tick_timeout_id);
      priv->tick_timeout_id = 0;
    }

  if (priv->uri != NULL)
    {
      g_free(priv->uri);
      priv->uri = NULL;
    }

  if (uri != NULL)
    {
      priv->uri = g_strdup(uri);
      if (priv->uri == NULL)
	return;

      vlc_media = libvlc_media_new(priv->vlc_instance,
				   priv->uri, &priv->vlc_exception);
      clutter_vlc_catch(&priv->vlc_exception);

      sprintf(media_arg, ":clutter-texture=%lld",
	      (long long int)(intptr_t)CLUTTER_TEXTURE(video_texture));

      libvlc_media_add_option(vlc_media, media_arg, &priv->vlc_exception);
      clutter_vlc_catch(&priv->vlc_exception);

      priv->vlc_media_player =
	libvlc_media_player_new_from_media(vlc_media,
					   &priv->vlc_exception);
      clutter_vlc_catch(&priv->vlc_exception);

      libvlc_media_release(vlc_media);

      priv->tick_timeout_id = g_timeout_add(TICK_TIMEOUT * 1000,
					    clutter_vlc_tick_timeout,
					    self);
    }

  g_object_notify(self, "uri");
  g_object_notify(self, "can-seek");
  g_object_notify(self, "duration");
  g_object_notify(self, "progress");
}


static void
clutter_vlc_set_playing(ClutterVlcVideoTexture* video_texture,
			gboolean playing)
{
  ClutterVlcVideoTexturePrivate* priv;

  priv = video_texture->priv;

  if (priv->vlc_instance == NULL)
    return;

  if (priv->uri == NULL)
    return;

  if (priv->vlc_media_player == NULL)
    return;

  if (playing == FALSE)
    {
      libvlc_media_player_pause(priv->vlc_media_player,
				&priv->vlc_exception);
    }
  else
    {
      libvlc_media_player_play(priv->vlc_media_player,
			       &priv->vlc_exception);
    }
  clutter_vlc_catch(&priv->vlc_exception);

  g_object_notify(G_OBJECT(video_texture), "playing");
  g_object_notify(G_OBJECT(video_texture), "progress");
}


static gboolean
clutter_vlc_get_playing(ClutterVlcVideoTexture* video_texture)
{
  ClutterVlcVideoTexturePrivate* priv;
  libvlc_state_t vlc_state;

  priv = video_texture->priv;

  if (priv->vlc_media_player == NULL)
    return FALSE;

  vlc_state = libvlc_media_player_get_state(priv->vlc_media_player,
					    &priv->vlc_exception);
  clutter_vlc_catch(&priv->vlc_exception);

  if (vlc_state != libvlc_Playing)
    return FALSE;

  return TRUE;
}


static void
clutter_vlc_set_progress(ClutterVlcVideoTexture* video_texture,
			 gdouble progress)
{
  ClutterVlcVideoTexturePrivate* priv;

  priv = video_texture->priv;

  if (priv->vlc_media_player == NULL)
    return;

  libvlc_media_player_set_position(priv->vlc_media_player, (float)progress,
				   &priv->vlc_exception);
  clutter_vlc_catch(&priv->vlc_exception);

  g_object_notify(G_OBJECT(video_texture), "progress");
}


static gdouble
clutter_vlc_get_progress(ClutterVlcVideoTexture* video_texture)
{
  ClutterVlcVideoTexturePrivate* priv;
  gdouble position;

  priv = video_texture->priv;

  if (priv->vlc_media_player == NULL)
    return 0.0;

  position = libvlc_media_player_get_position(priv->vlc_media_player,
					      &priv->vlc_exception);
  clutter_vlc_catch(&priv->vlc_exception);

  return position;
}


static void
clutter_vlc_set_audio_volume(ClutterVlcVideoTexture* video_texture,
			     gdouble volume)
{
  ClutterVlcVideoTexturePrivate* priv;

  priv = video_texture->priv;

  if (priv->vlc_instance == NULL)
    return;

  libvlc_audio_set_volume(priv->vlc_instance,
			  (int)(volume * 200.0),
			  &priv->vlc_exception);
  clutter_vlc_catch(&priv->vlc_exception);

  g_object_notify (G_OBJECT (video_texture), "audio-volume");
}


static gdouble
clutter_vlc_get_audio_volume(ClutterVlcVideoTexture* video_texture)
{
  ClutterVlcVideoTexturePrivate* priv;
  gdouble volume;

  priv = video_texture->priv;

  if (priv->vlc_instance == NULL)
    return 0.0;

  volume = (gdouble)libvlc_audio_get_volume(priv->vlc_instance,
					    &priv->vlc_exception);
  clutter_vlc_catch(&priv->vlc_exception);

  return CLAMP(volume / 200.0, 0.0, 1.0);
}


static gboolean
clutter_vlc_is_seekable(ClutterVlcVideoTexture* video_texture)
{
  ClutterVlcVideoTexturePrivate* priv;
  int ret;

  priv = video_texture->priv;

  if (priv->vlc_media_player == NULL)
    return FALSE;

  ret = libvlc_media_player_is_seekable(priv->vlc_media_player,
					&priv->vlc_exception);
  clutter_vlc_catch(&priv->vlc_exception);

  if (ret == 0)
    return FALSE;

  return TRUE;
}


static int
clutter_vlc_get_duration(ClutterVlcVideoTexture* video_texture)
{
  ClutterVlcVideoTexturePrivate* priv;
  libvlc_time_t ret;

  priv = video_texture->priv;

  if (priv->vlc_media_player == NULL)
    return 0;

  ret = libvlc_media_player_get_length(priv->vlc_media_player,
				       &priv->vlc_exception);
  clutter_vlc_catch(&priv->vlc_exception);

  return (int)ret;
}


static void
clutter_vlc_media_init(ClutterMediaIface* iface)
{

}

static void
clutter_vlc_video_texture_dispose(GObject* object)
{
  ClutterVlcVideoTexture* self;
  ClutterVlcVideoTexturePrivate* priv;

  self = CLUTTER_VLC_VIDEO_TEXTURE(object);
  priv = self->priv;

  if (priv->vlc_media_player != NULL)
    {
      libvlc_media_player_stop(priv->vlc_media_player,
			       &priv->vlc_exception);
      clutter_vlc_catch(&priv->vlc_exception);

      libvlc_media_player_release(priv->vlc_media_player);
      priv->vlc_media_player = NULL;

      g_source_remove(priv->tick_timeout_id);
      priv->tick_timeout_id = 0;
    }

  if (priv->vlc_instance != NULL)
    {
      libvlc_release(priv->vlc_instance);
      priv->vlc_instance = NULL;
    }

  G_OBJECT_CLASS(clutter_vlc_video_texture_parent_class)->dispose(object);
}


static void
clutter_vlc_video_texture_finalize(GObject* object)
{
  ClutterVlcVideoTexture* self;
  ClutterVlcVideoTexturePrivate* priv;

  self = CLUTTER_VLC_VIDEO_TEXTURE(object);
  priv = self->priv;

  if (priv->uri != NULL)
    {
      g_free(priv->uri);
      priv->uri = NULL;
    }

  G_OBJECT_CLASS(clutter_vlc_video_texture_parent_class)->finalize(object);
}


static void
clutter_vlc_video_texture_set_property(GObject* object,
				       guint property_id,
				       const GValue* value,
				       GParamSpec* pspec)
{
  ClutterVlcVideoTexture* video_texture;

  video_texture = CLUTTER_VLC_VIDEO_TEXTURE(object);

  switch (property_id)
    {
    case PROP_URI:
      clutter_vlc_set_uri(video_texture, g_value_get_string(value));
      break;

    case PROP_PLAYING:
      clutter_vlc_set_playing(video_texture, g_value_get_boolean(value));
      break;

    case PROP_PROGRESS:
      clutter_vlc_set_progress(video_texture, g_value_get_double(value));
      break;

    case PROP_AUDIO_VOLUME:
      clutter_vlc_set_audio_volume(video_texture, g_value_get_double(value));
      break;

    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID(object, property_id, pspec);
    }
}


static void
clutter_vlc_video_texture_get_property(GObject* object,
				       guint property_id,
				       GValue* value,
				       GParamSpec* pspec)
{
  ClutterVlcVideoTexture* video_texture;
  ClutterVlcVideoTexturePrivate* priv;

  video_texture = CLUTTER_VLC_VIDEO_TEXTURE(object);
  priv = video_texture->priv;

  switch (property_id)
    {
    case PROP_URI:
      g_value_set_string(value, priv->uri);
      break;

    case PROP_PLAYING:
      g_value_set_boolean(value, clutter_vlc_get_playing(video_texture));
      break;

    case PROP_PROGRESS:
      g_value_set_double(value, clutter_vlc_get_progress(video_texture));
      break;

    case PROP_AUDIO_VOLUME:
      g_value_set_double(value, clutter_vlc_get_audio_volume(video_texture));
      break;

    case PROP_CAN_SEEK:
      g_value_set_boolean(value, clutter_vlc_is_seekable(video_texture));
      break;

/*     case PROP_BUFFER_FILL: */
/*       g_value_set_double(value, priv->buffer_fill); */
/*       break; */

    case PROP_DURATION:
      g_value_set_int(value, clutter_vlc_get_duration(video_texture));
      break;

    default:
      G_OBJECT_WARN_INVALID_PROPERTY_ID(object, property_id, pspec);
    }
}


static void
clutter_vlc_video_texture_class_init(ClutterVlcVideoTextureClass* klass)
{
  GObjectClass* object_class;

  object_class = G_OBJECT_CLASS(klass);

  object_class->dispose = clutter_vlc_video_texture_dispose;
  object_class->finalize = clutter_vlc_video_texture_finalize;
  object_class->set_property = clutter_vlc_video_texture_set_property;
  object_class->get_property = clutter_vlc_video_texture_get_property;

  g_object_class_override_property(object_class,
				   PROP_URI, "uri");
  g_object_class_override_property(object_class,
				   PROP_PLAYING, "playing");
  g_object_class_override_property(object_class,
				   PROP_PROGRESS, "progress");
  g_object_class_override_property(object_class,
				   PROP_AUDIO_VOLUME, "audio-volume");
  g_object_class_override_property(object_class,
				   PROP_CAN_SEEK, "can-seek");
  g_object_class_override_property(object_class,
				   PROP_DURATION, "duration");
  g_object_class_override_property(object_class,
				   PROP_BUFFER_FILL, "buffer-fill");

  g_type_class_add_private(klass, sizeof(ClutterVlcVideoTexturePrivate));
}


static void
clutter_vlc_video_texture_init(ClutterVlcVideoTexture* video_texture)
{
  ClutterVlcVideoTexturePrivate* priv;
  char const * vlc_argv[] =
    {
      "--ignore-config",
      "--vout", "clutter",
    };
  int vlc_argc = sizeof(vlc_argv) / sizeof(*vlc_argv);

  video_texture->priv = CLUTTER_VLC_VIDEO_TEXTURE_GET_PRIVATE(video_texture);
  priv = video_texture->priv;

  libvlc_exception_init(&priv->vlc_exception);
  priv->vlc_instance = libvlc_new(vlc_argc, vlc_argv, &priv->vlc_exception);
  clutter_vlc_catch(&priv->vlc_exception);

  priv->vlc_media_player = NULL;
  priv->uri = NULL;
}


ClutterActor *
clutter_vlc_video_texture_new(void)
{
  return g_object_new(CLUTTER_VLC_TYPE_VIDEO_TEXTURE,
		      NULL);
}
